keymanager/src/churp/handler.rs

```diff
@@ -0,0 +1,263 @@
+
+        // Prepare response and sign it with RAK.
+        let application = ApplicationRequest {
+            id: churp_id,
+            runtime_id: self.runtime_id,
+            round,
+            checksum,
+        };
+        let body = cbor::to_vec(application.clone());
+        let signature = self
+            .signer
+            .sign(APPLICATION_REQUEST_SIGNATURE_CONTEXT, &body)?;
+
+        Ok(SignedApplicationRequest {
+            application,
+            signature,
+        })
+    }
+
+    fn get_or_create_dealer<D>(
+        &self,
+        churp_id: u8,
+        round: u64,
+        threshold: u8,
+    ) -> Result<Arc<Dealer<D>>>
+    where
+        D: DealerParams + 'static,
+    {
+        // Check the memory first.
+        let key = (self.runtime_id, round);
+        let mut dealers = self.dealers.write().unwrap();
+
+        if let Some(dealer) = dealers.get(&key) {
+            // Downcasting should never fail because the consensus ensures that
+            // the group ID cannot change.
+            let dealer = dealer
+                .downcast_ref::<Arc<Dealer<D>>>()
+                .ok_or(Error::DealerMismatch)?;
+            return Ok(dealer.clone());
+        }
+
+        // Check the local storage to ensure that only one secret bivariate
+        // polynomial is generated per round upon restarts, unless a malicious
+        // host has cleared the storage.
+        let polynomial = self
+            .storage
+            .load_bivariate_polynomial::<D::PrimeField>(churp_id, round)?;
+
+        let dealer = match polynomial {
+            Some(bp) => {
+                // Polynomial verification is redundant as encryption prevents
+                // tampering, while consensus ensures that the group ID remains
+                // unchanged and that polynomial dimensions remain consistent
+                // for any given pair of churp ID and round.
+                Dealer::new(bp)
+            }
+            None => {
+                // The local storage is empty. It's time to prepare a new polynomial.
+                // If the host has cleared the storage, other participants will detect
+                // the polynomial change because the checksum of the verification matrix
+                // in the submitted application will also change.
+                let dx = threshold.saturating_sub(1);
+                let dy = 2 * dx;
+
+                let dealer = match round {
+                    0 => Dealer::random(dx, dy, &mut OsRng),
+                    _ => Dealer::zero_hole(dx, dy, &mut OsRng),
+                };
+
+                // Encrypt and store the polynomial in case of a restart.
+                let polynomial = dealer.bivariate_polynomial();
+                self.storage
+                    .store_bivariate_polynomial(polynomial, churp_id, round)?;
+
+                dealer
+
+                // TODO: Delete previous polynomials. How? Should we call
```

Comment:
Currently there is no API for removals. We could add one if needed.

Reply:
Didn't know that, actually. Will change this so that we store only 2 values, one for the current round and one for the next one.

Reply:
Actually, one is enough. Will change latter if needed.

Reply:
In that case, just remove the comment.